lib/std/fs/path.zig

```diff
@@ -76,67 +77,115 @@ fn joinSep(allocator: *Allocator, separator: u8, sepPredicate: fn (u8) bool, pat
         buf_index += adjusted_path.len;
     }
 
+    if (zero) buf[buf.len - 1] = 0;
+
     // No need for shrink since buf is exactly the correct size.
     return buf;
 }
 
 /// Naively combines a series of paths with the native path seperator.
 /// Allocates memory for the result, which must be freed by the caller.
 pub fn join(allocator: *Allocator, paths: []const []const u8) ![]u8 {
-    return joinSep(allocator, sep, isSep, paths);
+    return joinSepMaybeZ(allocator, sep, isSep, paths, false);
 }
 
-fn testJoinWindows(paths: []const []const u8, expected: []const u8) void {
+/// Naively combines a series of paths with the native path seperator and null terminator.
+/// Allocates memory for the result, which must be freed by the caller.
+pub fn joinZ(allocator: *Allocator, paths: []const []const u8) ![:0]u8 {
+    const out = joinSepMaybeZ(allocator, sep, isSep, paths, true);
+    return out[0 .. out.len - 1 :0];
+}
+
+fn testJoinMaybeZWindows(paths: []const []const u8, expected: []const u8, zero: bool) void {
     const windowsIsSep = struct {
         fn isSep(byte: u8) bool {
             return byte == '/' or byte == '\\';
         }
     }.isSep;
-    const actual = joinSep(testing.allocator, sep_windows, windowsIsSep, paths) catch @panic("fail");
+    const actual = joinSepMaybeZ(testing.allocator, sep_windows, windowsIsSep, paths, zero) catch @panic("fail");
     defer testing.allocator.free(actual);
-    testing.expectEqualSlices(u8, expected, actual);
+    testing.expectEqualSlices(u8, expected, if (zero) actual[0 .. actual.len - 1 :0] else actual);
 }
 
-fn testJoinPosix(paths: []const []const u8, expected: []const u8) void {
+fn testJoinMaybeZPosix(paths: []const []const u8, expected: []const u8, zero: bool) void {
     const posixIsSep = struct {
         fn isSep(byte: u8) bool {
             return byte == '/';
         }
     }.isSep;
-    const actual = joinSep(testing.allocator, sep_posix, posixIsSep, paths) catch @panic("fail");
+    const actual = joinSepMaybeZ(testing.allocator, sep_posix, posixIsSep, paths, zero) catch @panic("fail");
     defer testing.allocator.free(actual);
-    testing.expectEqualSlices(u8, expected, actual);
+    testing.expectEqualSlices(u8, expected, if (zero) actual[0 .. actual.len - 1 :0] else actual);
 }
 
 test "join" {
-    testJoinWindows(&[_][]const u8{ "c:\\a\\b", "c" }, "c:\\a\\b\\c");
-    testJoinWindows(&[_][]const u8{ "c:\\a\\b", "c" }, "c:\\a\\b\\c");
-    testJoinWindows(&[_][]const u8{ "c:\\a\\b\\", "c" }, "c:\\a\\b\\c");
+    testJoinMaybeZWindows(&[_][]const u8{}, "", false);
```

Comment:
You might as well combine the two with:
```zig
for (&[_]bool{false, true}) |zero| {
    testJoinMaybeZWindows(&[_][]const u8{}, "", zero);
    ......
```

Reply:
I noticed the duplication here but not sure if we want to merge them into a single test case. Like mem.join and mem.joinZ, they're testing different behaviours. So what's the common practice for testing Z and non-Z functions?